### DISCUSSION TITLE
Sharing a Pluto webserver from outside Julia

### DISCUSSION THREAD

**Discussion by heetbeet** [fonsp/Pluto.jl#1161] [Pluto development] [answered]
In Juliawin, we provide the following executable, essentially running the this block of code, for starting a pluto server:
![image](https://user-images.githubusercontent.com/4103775/118243401-b3d75c00-b49e-11eb-926f-3e1c46a5411c.png)

But every time you run pluto.exe, it creates a new instance of Pluto, and then open up a browser tab to access that instance. I want something that tries (by default) to connect to current running Pluto instances. 

I have the following idea. If no arguments are provided to pluto.exe, run Pluto in a kind of "shared-instance-mode". I can think of the following way to create such a mode:
- Look in the file userdata/plutosecret if a previous secret exist
- Try pinging the usual url+secret pattern to see if Pluto is already running
- Either connect to this already running Pluto if found, or create a new secret and new instance

What worries me is the fact that the secret would be out in the open (on a file on the harddrive), but it will still only be accessible from the harddrive of the running Julia instance. So in my mind the security is still close to running a general Pluto instance, since an external computer still cannot access the secret. But I don't want to be wrong.

1. Are there better ways to share a Pluto instance between processes running external to the current Julia instance?
2. How do I "ping" a Pluto url to see if it is currently serving Pluto?

**Comment by fonsp** [comment, marked as answer]
1. Seems like a more general process management question, which I don't know anything about. Maybe someone else can help, or you can ask the more general (not Pluto-specific) question on stackoverflow.com.
2. The simplest way is to check if the address returns a 200 response. There is also `http://localhost:1234/ping`. Are you looking for something more specific?

**Comment by fonsp** [comment]
@heetbeet Did you find an answer to your question already? Did my comment help?

**Reply by heetbeet** [reply]
Hi, sorry for no feedback, I've not taken another crack at the problem, yet. 
(2.) this is exactly the type of thing I'm looking for, will just have to test what happens if another app hogs the port I'm pinging (or is this ping functionality Pluto specific?). I think someone mentioned somewhere that a certain pluto url returns with the text "ok" I'll also try that approach
(1.) I think I'm just going to implement the feature as it is in my head on a side branch, and then ask around on SO Slack etc. to see how bad of a choice it is (or not), and then reiterate

I'll post my findings and updates here. First priority for that project is to build a proper release/installer creator, and then finishing up my ubuntu port... then I'll focus again on better Pluto integration

**Reply by fonsp** [reply]
Search for `ping` in https://github.com/fonsp/Pluto.jl/blob/main/src/webserver/Static.jl

**Reply by heetbeet** [reply]
Thanks!

**Comment by heetbeet** [comment]
I have settled on the following approach for sniffing out current Pluto runner:
https://github.com/heetbeet/juliawin/blob/main/bin/(pluto.bat)

Basically, 

- it stores the secret in a file
- the next run grabs the secret, then sniffs for any Pluto servers
- if a Pluto server is running with this secret, then it latches on to that server
- if no Pluto server is running with this secret, then the process is reset
